src/Assets/Api.php

```diff
@@ -84,6 +84,14 @@ public function register_script( $handle, $relative_src, $dependencies = [], $ha
 			$version = $this->get_file_version( $relative_src );
 		}
 
+		if ( in_array( $handle, $dependencies, true ) ) {
+			$dependencies = array_diff( $dependencies, [ $handle ] );
+			if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
+				// phpcs:ignore
+				trigger_error( sprintf( 'Script with handle %s had a dependency on itself which we removed. This is an indicator that your JS code has a circular dependency that can cause bugs.', $handle ), E_USER_WARNING );
```

Comment:
If the dependency is removed automatically, I wonder if it makes more sense avoiding the pronoun ("has been removed" instead of "we removed").
```suggestion
				trigger_error( sprintf( 'Script with handle %s had a dependency on itself which has been removed. This is an indicator that your JS code has a circular dependency that can cause bugs.', $handle ), E_USER_WARNING );
```